components/script/dom/bindings/str.rs

```diff
@@ -332,3 +346,74 @@ impl Extend<char> for DOMString {
         self.0.extend(iterable)
     }
 }
+
+/// https://html.spec.whatwg.org/multipage/#parse-a-month-component
+fn parse_month_component(value: &str) -> Result<(u32, u32), ()> {
+    // Step 3
+    let value_vec: Vec<&str> = value.split('-').collect();
+    if value_vec.len() < 2 {
+        return Err(());
+    }
+    let year = value_vec[0];
+    let month = value_vec[1];
+
+    // Step 1, 2
+    if year.len() < 4 ||
+        !year.chars().all(|c| c.is_digit(10)) ||
+        year == "0000" {
+        return Err(());
+    }
+
+    // Step 4, 5
+    let month_int = month.parse::<u32>().map_err(|_| ())?;
+    if month.len() != 2 ||  month_int > 12 || month_int < 1 {
+        return Err(());
+    }
+
+    // Step 6
+    let year_int = year.parse::<u32>().map_err(|_| ())?;
+    Ok((year_int, month_int))
+}
+
+/// https://html.spec.whatwg.org/multipage/#parse-a-date-component
+fn parse_date_component(value: &str) -> Result<(u32, u32, u32), ()> {
+    // Step 3
+    let value_vec: Vec<&str> = value.split('-').collect();
```

Comment:
Same comment for validating month strings -- we want to reduce the amount of heap allocations.